keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -207,82 +257,114 @@ resources allocated (Pod.Spec.Containers[i].ResourcesAllocated) for all Pods in
 the Node, except the Pod being resized. For the Pod being resized, it adds the
 new desired resources (i.e Spec.Containers[i].Resources.Requests) to the sum.
 * If new desired resources fit, Kubelet accepts the resize by updating
-  Pod.Spec.Containers[i].ResourcesAllocated, and then proceeds to invoke
-  UpdateContainerResources CRI API to update Container resource limits. Once
-  all Containers are successfully updated, it updates
-  Pod.Status.ContainerStatuses[i].Resources to reflect new resource values.
-* If new desired resources don't fit, Kubelet rejects the resize, and no
-  further action is taken.
-  - Kubelet retries the Pod resize at a later time.
-
-If multiple Pods need resizing, they are handled sequentially in the order in
-which Pod additions and updates arrive at Kubelet.
+  Status...ResourcesAllocated field and setting Status...Resize to
+  "InProgress".  It then invokes the UpdateContainerResources CRI API to update
+  Container resource limits.  Once all Containers are successfully updated, it
+  updates Status...Resources to reflect new resource values and unsets
+  Status...Resize.
+* If new desired resources don't fit, Kubelet will update the Status...Resize
+  field to "Infeasible" and does not act on the resize.
+
+In the future, this may evolve to support "deferred" resizes, wherein the
+kubelet will try again to admit the resized pod.  For alpha, at least, this is
+not required.  If this functionality is desired, a new value for
+Status...Resize will be needed.
+
+<<[UNRESOLVED]>>
+If we don't have kubelet retry, then there's no real way for a VPA-like actor
```

Comment:
We DO have retry in kubelet. We keep retrying as long as Spec.C[i].Resources != Status.CS[i].Resources. (Or Resize==Infeasible with addition of this new summarized signal field).

Currently, we retry periodically in kubelet's syncPod function for simplicity. In the future, we may make it smarter to try based on events such as other pods departing (and thus potentially creating room for resize). We don't have a way for VPA-like actor control the retry frequency. They certainly can try an end-run via API subject to apiserver rate limiting.

Reply:
Maybe I missed this before.  What circumstances would lead to a retry vs deciding it is Infeasible?

In my mind this comes back to the question: "How does the VPA-like actor know what is happening".  I was assuming kubelet would be simple and IMMEDIATELY decide "this is feasible" (resize = InProgress) or "this is not feasible" (resize = Infeasible).  I also assumed that Infeasible is a terminal state.

If we're willing to say "this is feasible in theory, but not right now" then I think we'd want a different value for resize.

I'm totally OK with this, I left it out for simplicity.  If you are set on that, I'll add it back.  In fact, I will add another commit to show.  If you don't like it we can abandon that.

Reply:
Today, for simplicity, we always retry i.e periodically call canAdmitPod(newDesiredResources) function periodically as long as Spec.C[i].Resources != Status.CS[i].Resources, regardless of spec resources can be accommodated maybe-later (not enough room on the node at this time - may be accommodated at later time when other pods leave) or never (request exceeds node capacity - can never be accommodated).

With introduction of Deferred & Infeasible, we can perform retry if Status.Resize == Deferred, and not attempt it if Status.Resize ==  Infeasible, because we know it can never be accommodated. If status were lost, this can be reconstructed.